### ISSUE TITLE
Add reconciliation API

### ISSUE THREAD

**Issue by awagner-mainz** [hbz/skohub-pubsub#54]
Hi all,

I would like to use my SKOS taxonomy in OpenRefine or TEI Publisher for reconciliation and entity annotation. (And I can imagine this could be a common use case.) For this, it would be ideal if the skohub.io server exposed a [Reconciliation API](https://reconciliation-api.github.io/specs/latest/) in addition to the Activity Streams API. Since skohub-pubsub is already providing a web service (other than, say, skohub-vocabs), I am opening this as an issue in this project, even though it is obviously a different thing than subscription to a vocabulary.

The way I see it is this: we'd need (a) a frontend interface (say in express.js) that would transform some search server's responses to proper Reconciliation API responses. Since skohub-pubsub already depends on ElasticSearch, we could use just (b) another ElasticSearch index for this purpose. We could also (b') add a dependency on FlexSearch or even (b'') make the search server configurable (with a choice of suitable connector go-betweens). Either way, probably we'd set up the search server so that it (c) has separate (or separable) indexes per skohub "tenant"/project and that it can be populated both (d) "run-time" (with a REST call?), whenever new projects or vocabularies are added via the github-pubsub call and (e) "from scratch" when the server needs to be restarted (from the file system?).

For a minimal setup, we'd need to go with service manifest and query response objects. I will attach sample objects later. If there are no objections, I'll try my luck with a pipeline Elasticsearch query - Reconciliation objects - express.js API...

What do you think?

**Comment by acka47** [comment]
Thanks for the detailed request, @awagner-mainz . I agree that this use case should be supported by SkoHub. Here are some thoughts about it:

- You are correct that an Elasticsearch index is already used in SkoHub PubSub. However, I don't think we should implement this feature as part of PubSub. It would fit better into https://github.com/skohub-io/skohub-vocabs or maybe even an additional module would make sense. Until we have sorted this out, we can continue the discussion here.
- I am not sure whether we should make this a reconciliation endpoint optional as you probably don't need reconciliation for small value lists which would mean that the implementation would ideally go hand in hand with the possibility to configure the publication of a vocab (which is also necessary for other things, like https://github.com/skohub-io/skohub-vocabs/issues/75 ). On the other hand, you could also use the [suggest service](https://reconciliation-api.github.io/specs/latest/#suggest-services) for lookups.

**Comment by awagner-mainz** [comment]
> * You are correct that an Elasticsearch index is already used in SkoHub PubSub. However, I don't think we should implement this feature as part of PubSub. It would fit better into https://github.com/skohub-io/skohub-vocabs or maybe even an additional module would make sense.

I agree that this is possibly best thought of as an additional, independent module. But knowing that there is also the pubsub service (and thus an Elasticsearch instance) running on skohub.io, we could prioritize Eleasticsearch or at least define it as a backend that we definitely will want to support. And it should not hurt ES if it is running indexes for both pubsub and reconciliation APIs, should it? If we do so, we have to decide whether we want to be able to switch backends (and add, say, a FlexSearch server as well) or just hard-wire everything for one backend.

**Comment by literarymachine** [comment]
I agree with @acka47 that this would make more sense in `skohub-vocabs` or as an extra module. The latter is probably cleaner, but most likely also more work. If we integrate this in `skohub-vocabs`, we could try to reuse the flexsearch-indexes that are already being built, on the server side. We would need a small node.js application that transforms reconciliation requests to flexsearch queries and then flexsearch results back to reconcilitaion responses. The hardest part is probably to come up with a proper scoring.

**Comment by awagner-mainz** [comment]
Another component that just came to my mind is routing: I assume we would rather want to offer one endpoint https://skohub.io/reconcile only and use the `properties` key in the query to differentiate between "tenant"/projects rather than offer https://skohub.io/rg-mpg-de/vocabs-polmat/heads/main/w3id.org/rg-mpg-de/polmat/reconcile and https://skohub.io/hbz/vocabs-edu/heads/master/w3id.org/class/esc/reconcile (and so on) as different endpoints. If so, we don't need a routing component. Just to have this for the record.

**Comment by acka47** [comment]
> If we integrate this in skohub-vocabs, we could try to reuse the flexsearch-indexes that are already being built, on the server side.

As we have a lot of experience with providing a reconciliation endpoint based on Elasticsearch for GND (https://lobid.org/gnd/reconcile) this might be the easier route for us.

>  I assume we would rather want to offer one endpoint https://skohub.io/reconcile only and use the `properties` key in the query to differentiate between "tenant"/projects rather than offer https://skohub.io/rg-mpg-de/vocabs-polmat/heads/main/w3id.org/rg-mpg-de/polmat/reconcile and https://skohub.io/hbz/vocabs-edu/heads/master/w3id.org/class/esc/reconcile (and so on) as different endpoints.

I think we will have to offer one endpoint per vocab and might also provide it via `https://skohub.io/reconcile?vocab=w3id.org/rg-mpg-de/polmat/scheme` or similar. At least the [identifier space](https://reconciliation-api.github.io/specs/latest/#identifier-and-schema-spaces) is different per vocab (the schema space being identical for all SkoHub vocabs) and a reconciliation endpoint can only declare one identifier space.

With regard to using the `properties` key to differentiate between vocabs, I guess you are thinking about sending along something like below, aren't you? I am afraid that this wouldn't be very user friendly compared to one reconciliation endpoint per vocab.

```json
{
  "properties":[
    {
      "pid":"inScheme",
      "v":[
        "https://w3id.org/rg-mpg-de/polmat/scheme"
      ]
    }
  ]
}
```

**Comment by awagner-mainz** [comment]
> As we have a lot of experience with providing a reconciliation endpoint based on Elasticsearch for GND (https://lobid.org/gnd/reconcile) this might be the easier route for us.

@acka47 , could you perhaps share a sample ES query as it's being called by lobid's reconciliation API middleware? and perhaps lobid's ES document definition?

**Comment by awagner-mainz** [comment]
> [something about one single endpoint for the service vs. individual endpoints for each vocab.]
>
> With regard to using the `properties` key to differentiate between vocabs, I guess you are thinking about sending along something like below, aren't you? I am afraid that this wouldn't be very user friendly compared to one reconciliation endpoint per vocab.

Yes, in terms of user friendliness I agree 100%. And, coming to think about it, probably good user experience should be prioritized over easy implementation ("LOUD data").

**Comment by acka47** [comment]
> @acka47 , could you perhaps share a sample ES query as it's being called by lobid's reconciliation API middleware? and perhaps lobid's ES document definition?

Here is the implementation for lobid-gnd: https://github.com/hbz/lobid-gnd/blob/master/app/controllers/Reconcile.java The index setting is here: https://github.com/hbz/lobid-gnd/blob/master/conf/index-settings.json We don't have a JSON Schema for lobid-gnd data, though. Pinging @fsteeg as he is the one who implemented this.

**Comment by awagner-mainz** [comment]
If I should (and I want to) give a first try to develop an additional module for this issue, it would obviously be good if it were as analogous to the other modules as possible, in terms of code organisation and libraries/dependencies. Now I found that vocabs uses koa as a web framework whereas pubsub uses express. Being the newcomer that I am, may I ask why you chose differently for the two modules? I hope this does not clutter the issue and that the question is not a stupid one.

**Comment by fsteeg** [comment]
> could you perhaps share a sample ES query as it's being called by lobid's reconciliation API

This is what an ES reconciliation query basically looks like in lobid-gnd, with the main name query in `must`, a specified type in `filter` and one additional property in `should` (I removed some GND-specific stuff and details to make it more readable):

```
{
  "query": {
    "bool": {
      "must": [
        {
          "query_string": {
            "query": "Twain, Mark",
            "fields": [
              "preferredName^4.0",
              "temporaryName^1.0",
              "variantName^2.0"
            ]
          }
        }
      ],
      "filter": [
        {
          "query_string": {
            "query": "type:DifferentiatedPerson"
          }
        }
      ],
      "should": [
        {
          "query_string": {
            "query": "professionOrOccupationAsLiteral:Schriftsteller",
            "boost": 5.0
          }
        }
      ]
    }
  }
}
```

I hope that helps. For details, we're building the ES reconciliation queries here: https://github.com/hbz/lobid-gnd/blob/c0cd807110f115f2c722124ab3b65c2c4cfe52f6/app/controllers/Reconcile.java#L452

**Comment by fsteeg** [comment]
> I think we will have to offer one endpoint per vocab and might also provide it via `https://skohub.io/reconcile?vocab=w3id.org/rg-mpg-de/polmat/scheme` or similar.

Could it make sense to use the vocabs as `type`? With the `defaultTypes` (and maybe a type suggest service) this could provide a nice way to discover available vocabs in OpenRefine.

**Comment by awagner-mainz** [comment]
Coming back to this tonite after other business had to be looked after in the past days, I think it could be thus:

1. Create a new tool, working title "skohub-reconc" or so, that translates HTTP API requests from the web to queries to a search engine and (transforms and) ships the results as reconciliation API responses. This would obviously depend on a search engine like Elasticsearch (or Flexsearch, but I, for one, have some experience with Elastic only).
2. When configuring skohub-vocabs, specify if a Reconciliation API should be running. If so, in addition to creating html and json files, it pushes a vocabulary's data to the search engine upon rendering.

While we could have skohub-vocabs push documents to the search engine directly, I would think it would perhaps be "cleaner" to push it via the skohub-reconc service. Then this service would proxy/shield everything from the search server, be responsible for CRUD operations on the reconc. info database, and could provide interfaces to several different backends without skohub-vocabs having to care about it at all. Also, re-populating the search index from scratch in case of data loss could be done with a special call to the skohub-reconc tool, and it could work based on what is in the filesystem (rather than on receiving data packages the triggering of which would probably become quite tricky - I imagine a scenario where the search index is lost but the vocabs files are still present. Then, we wouldn't need to rebuild all the vocabularies just for the reconc search index. I have no idea how likely or absurd such a scenario is.).

I suppose routing, types and some other things will fall into place once the general approach is agreed upon.

**Comment by awagner-mainz** [comment]
just a very short notice about the progress I have made: I have created a repository with an express app that serves the API and has elasticsearch code infrastructure in place. Next, I need to test (and probably adjust a lot of things) and then find a way to populate it from skohub-vocab runs. Repository is here: <https://github.com/rg-mpg-de/skohub-reconcile>

**Comment by awagner-mainz** [comment]
I have a working prototype! Only handling the happy path at the moment, no schema or service validation and no test coverage, but it works:

1. At <https://github.com/rg-mpg-de/skohub-reconcile> there is another component running as an Reconciliation Service over an elasticsearch server. (Start with `npm start`, this will connect to ES and delete and re-initialize a 'skohub-reconcile' index. It handles reconciliation queries only, no suggestions or other things at the moment.)
2. At <https://github.com/rg-mpg-de/skohub-vocabs/tree/feature-reconc> there is a fork of skohub-vocabs that has added a script that can be launched with `npm run populate-reconc`. At the moment this takes a hardcoded turtle file from the data folder, creates documents suitable for the ES mapping and sends them to bulk-indexing.

After these two steps, it is possible to launch reconciliation queries against the skohub-reconcile service. I am using [insomnia](https://insomnia.rest/) for REST debugging and have exported some queries I have been using. I have attached them to this message.

[insomnia_2021-12-04.json.zip](https://github.com/skohub-io/skohub-pubsub/files/7653125/insomnia_2021-12-04.json.zip)

**Comment by fsteeg** [comment]
Sounds great @awagner-mainz! Looking forward to take a closer look this week.

**Comment by awagner-mainz** [comment]
I will probably need to improve how it is integrated with the other skohub modules: what version of elasticsearch they require respectively, how skohub-reconcile is launched, and how a ttl data file can be passed as to-be-processed. Also, at the moment, re-processing a ttl file duplicates entries in the elastic index rather than updating them (and deleting the elastic index on every restart of skohub-reconcile is probably not very clever either).

In short: many things that need to be addressed, but in principle it's working. We should collect these issues in [skohub-reconcile's issue tracker](https://github.com/rg-mpg-de/skohub-reconcile/issues) I'd think.

**Comment by fsteeg** [comment]
So I've got the basic setup working with the polmat.ttl data. I will open a PR for the README with some tweaks and things I had to do to make it work so far. I'll also open issues for things I noticed while testing.

**Comment by awagner-mainz** [comment]
Since we are tracking this in [skohub-reconcile](https://github.com/rg-mpg-de/skohub-reconcile/issues) (and (my fork's) [skohub-vocab](https://github.com/rg-mpg-de/skohub-vocabs/issues)) issues, I guess we can close this issue here...
